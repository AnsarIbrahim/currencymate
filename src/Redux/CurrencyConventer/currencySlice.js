import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

import BaseUrl from "../baseUrl.js";

export const fetchCurrency = createAsyncThunk(
  "currency/fetchCurrency",
  async () => {
    const response = await axios.get(BaseUrl);
    return response.data;
  }
);

const currencySlice = createSlice({
  name: "currency",
  initialState: {
    currency: {},
    status: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCurrency.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchCurrency.fulfilled, (state, { payload }) => {
        state.currency = payload;
        state.status = "success";
      })
      .addCase(fetchCurrency.rejected, (state) => {
        state.status = "failed";
      });
  },
});

export default currencySlice.reducer;
