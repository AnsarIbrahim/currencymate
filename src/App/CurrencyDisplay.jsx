import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import Loading from "../components/Loading";
import Error from "../components/Error";
import { fetchCurrency } from "../Redux/CurrencyConventer/currencySlice";

const CurrencyDisplay = () => {
  const dispatch = useDispatch();
  const currencyData = useSelector((state) => state.currency.currency);
  const status = useSelector((state) => state.currency.status);

  const [amount, setAmount] = useState(1);
  const [targetCurrency, setTargetCurrency] = useState("USD");

  useEffect(() => {
    dispatch(fetchCurrency());
  }, [dispatch]);

  const handleAmountChange = (e) => {
    setAmount(e.target.value);
  };

  const handleCurrencyChange = (e) => {
    setTargetCurrency(e.target.value);
  };

  if (status === "loading") {
    return <Loading />;
  } else if (status === "failed") {
    return <Error message="Failed to load currency data" />;
  }

  let conversionRate = 0;
  let convertedAmount = 0;

  if (currencyData && currencyData.conversion_rates) {
    conversionRate = currencyData.conversion_rates[targetCurrency];
    convertedAmount = (amount * conversionRate).toFixed(2);
  }

  return (
    <div className="flex flex-col items-center justify-center mt-2 space-y-5">
      <h1 className="text-3xl font-bold text-blue-600 mb-2 font-space-grotesk">
        Currency Converter
      </h1>
      <div className="bg-white shadow-lg rounded-lg p-5 w-80">
        <div className="flex space-x-5">
          <input
            type="number"
            value={amount}
            onChange={handleAmountChange}
            className="w-full border-2 border-blue-200 rounded-md p-2 shadow-md font-space-grotesk text-xs font-semibold"
            placeholder="Amount"
          />
          {currencyData && currencyData.conversion_rates && (
            <select
              value={targetCurrency}
              onChange={handleCurrencyChange}
              className="w-full border-2 border-blue-200 rounded-md p-2 shadow-md font-poppins text-xs font-semibold"
            >
              {Object.keys(currencyData.conversion_rates).map((currency) => (
                <option key={currency} value={currency}>
                  {currency}
                </option>
              ))}
            </select>
          )}
        </div>
        <div className="border-t border-blue-200 mt-4 pt-4 text-xs font-space-grotesk font-semibold">
          <span className="text-green-600">{amount}</span> USD is equivalent to{" "}
          <span className="text-red-400">{convertedAmount}</span>{" "}
          {targetCurrency}
        </div>
      </div>
    </div>
  );
};

export default CurrencyDisplay;
