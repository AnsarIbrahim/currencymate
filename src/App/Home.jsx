import React from "react";
import { useSelector } from "react-redux";

import CurrencyDisplay from "./CurrencyDisplay";
import ScrollingPrices from "../components/ScrollPrices";

const Home = () => {
  const conversionRates = useSelector(
    (state) => state.currency.currency.conversion_rates
  );

  const timeUpdate = useSelector(
    (state) => state.currency.currency.time_last_update_utc
  );

  return (
    <div className="min-h-screen bg-gradient-to-r from-blue-500 to-blue-100 flex flex-col items-center justify-center space-y-10 p-5">
      <h1 className="text-xl font-bold text-white font-poppins md:text-4xl">
        Currency<span className="md:text-xl text-blue-600 text-2xl">Mate</span>
      </h1>
      {conversionRates && (
        <ScrollingPrices conversionRates={conversionRates} direction="rtl" />
      )}
      <p className="text-white text-xs md:text-sm bg-blue-900 bg-opacity-50 rounded-xl p-2">
        Last updated:{" "}
        <span className="text-yellow-300">
          {new Date(timeUpdate).toLocaleString()}
        </span>
      </p>
      {conversionRates && (
        <ScrollingPrices conversionRates={conversionRates} direction="ltr" />
      )}
      <div className="bg-slate-300 shadow-lg rounded-lg p-5 w-full max-w-md pb-10">
        <CurrencyDisplay />
      </div>
    </div>
  );
};

export default Home;
