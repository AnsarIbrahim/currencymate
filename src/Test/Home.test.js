import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";

import currencySlice from "../Redux/CurrencyConventer/currencySlice";
import Home from "../App/Home";
import ScrollingPrices from "../components/ScrollPrices";
import CurrencyDisplay from "../App/CurrencyDisplay";

jest.mock("axios", () => ({
  get: jest.fn(() => Promise.resolve({ data: {} })),
}));

describe("Home component", () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        currency: currencySlice,
      },
      preloadedState: {
        currency: {
          currency: {
            conversion_rates: {
              USD: 1,
              EUR: 0.85,
            },
            time_last_update_utc: new Date().toISOString(),
          },
        },
      },
    });
  });

  it("renders Home component with title and last updated time", () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    expect(screen.getByText(/Currency/i)).toBeInTheDocument();
    expect(screen.getByText(/Mate/i)).toBeInTheDocument();
    expect(screen.getByText(/Last updated:/i)).toBeInTheDocument();
  });

  it("renders ScrollingPrices component", () => {
    const store = configureStore({
      reducer: {
        currency: currencySlice,
      },
      preloadedState: {
        currency: {
          currency: {
            conversion_rates: {
              USD: 1,
              EUR: 0.85,
            },
            time_last_update_utc: new Date().toISOString(),
          },
        },
      },
    });

    render(
      <Provider store={store}>
        <ScrollingPrices conversionRates={{ USD: 1, EUR: 0.85 }} />
      </Provider>
    );
  });

  it("renders CurrencyDisplay component", () => {
    const store = configureStore({
      reducer: {
        currency: currencySlice,
      },
      preloadedState: {
        currency: {
          currency: {
            conversion_rates: {
              USD: 1,
              EUR: 0.85,
            },
            time_last_update_utc: new Date().toISOString(),
          },
        },
      },
    });

    render(
      <Provider store={store}>
        <CurrencyDisplay />
      </Provider>
    );
  });

  it("matches snapshot for Home component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for ScrollingPrices component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ScrollingPrices conversionRates={{ USD: 1, EUR: 0.85 }} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for CurrencyDisplay component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <CurrencyDisplay />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
