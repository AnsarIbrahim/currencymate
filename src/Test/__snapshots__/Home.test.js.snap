// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component matches snapshot for CurrencyDisplay component 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center"
  >
    <span
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      style="display: inherit; position: relative; width: 60px; height: 60px;"
    >
      <span
        style="position: absolute; height: 60px; width: 60px; background-color: rgb(18, 58, 188); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 1s infinite ease-in-out;"
      />
      <span
        style="position: absolute; height: 60px; width: 60px; background-color: rgb(18, 58, 188); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 0s infinite ease-in-out;"
      />
    </span>
  </div>
</DocumentFragment>
`;

exports[`Home component matches snapshot for Home component 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen bg-gradient-to-r from-blue-500 to-blue-100 flex flex-col items-center justify-center space-y-10 p-5"
  >
    <h1
      class="text-xl font-bold text-white font-poppins md:text-4xl"
    >
      Currency
      <span
        class="md:text-xl text-blue-600 text-2xl"
      >
        Mate
      </span>
    </h1>
    <div
      class="w-[80%]"
    >
      <div
        class="sc-guDLey lcyjPN"
      >
        <div
          class="sc-dmyCSP AApyK"
          direction="rtl"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dmyCSP AApyK"
          direction="rtl"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dmyCSP AApyK"
          direction="rtl"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      class="text-white text-xs md:text-sm bg-blue-900 bg-opacity-50 rounded-xl p-2"
    >
      Last updated: 
      <span
        class="text-yellow-300"
      >
        23/04/2024, 21:53:55
      </span>
    </p>
    <div
      class="w-[80%]"
    >
      <div
        class="sc-guDLey lcyjPN"
      >
        <div
          class="sc-dmyCSP hKjfKh"
          direction="ltr"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dmyCSP hKjfKh"
          direction="ltr"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dmyCSP hKjfKh"
          direction="ltr"
          style="animation-duration: 0s;"
        >
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                USD:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                1.00
              </span>
            </div>
          </div>
          <div
            class="sc-beySPh bWwetx"
            position="relative"
            style="left: 0px; top: 0px;"
          >
            <div
              class="mx-2"
              style="white-space: nowrap;"
            >
              <span
                class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
              >
                EUR:
              </span>
              <span
                class="text-xs font-bold font-space-grotesk text-red-500"
              >
                0.85
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="bg-slate-300 shadow-lg rounded-lg p-5 w-full max-w-md pb-10"
    >
      <div
        class="flex justify-center items-center"
      >
        <span
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
          style="display: inherit; position: relative; width: 60px; height: 60px;"
        >
          <span
            style="position: absolute; height: 60px; width: 60px; background-color: rgb(18, 58, 188); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 1s infinite ease-in-out;"
          />
          <span
            style="position: absolute; height: 60px; width: 60px; background-color: rgb(18, 58, 188); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 0s infinite ease-in-out;"
          />
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home component matches snapshot for ScrollingPrices component 1`] = `
<DocumentFragment>
  <div
    class="w-[80%]"
  >
    <div
      class="sc-guDLey lcyjPN"
    >
      <div
        class="sc-dmyCSP AApyK"
        direction="rtl"
        style="animation-duration: 0s;"
      >
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              USD:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              1.00
            </span>
          </div>
        </div>
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              EUR:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              0.85
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-dmyCSP AApyK"
        direction="rtl"
        style="animation-duration: 0s;"
      >
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              USD:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              1.00
            </span>
          </div>
        </div>
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              EUR:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              0.85
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-dmyCSP AApyK"
        direction="rtl"
        style="animation-duration: 0s;"
      >
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              USD:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              1.00
            </span>
          </div>
        </div>
        <div
          class="sc-beySPh bWwetx"
          position="relative"
          style="left: 0px; top: 0px;"
        >
          <div
            class="mx-2"
            style="white-space: nowrap;"
          >
            <span
              class="text-xs font-bold font-space-grotesk text-green-700 mx-1"
            >
              EUR:
            </span>
            <span
              class="text-xs font-bold font-space-grotesk text-red-500"
            >
              0.85
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
